src/util/mapbox.js

```diff
@@ -119,3 +126,108 @@ function formatUrl(obj: UrlObject): string {
     const params = obj.params.length ? `?${obj.params.join('&')}` : '';
     return `${obj.protocol}://${obj.authority}${obj.path}${params}`;
 }
+
+class TurnstileEvent {
+    STORAGE_TOKEN: string;
+    localStorageAvailable: boolean;
+    eventData: Object;
+    queue: Array<number>;
+    pending: boolean
+    pendingRequest: ?Cancelable;
+
+    constructor() {
+        this.STORAGE_TOKEN = 'mapbox.turnstileEventData';
+        this.localStorageAvailable = storageAvailable('localStorage');
+        this.eventData = { anonId: null, lastSuccess: null };
+        this.queue = [];
+        this.pending = false;
+        this.pendingRequest = null;
+    }
+
+    queueRequest(date: number) {
+        this.queue.push(date);
+        this.processRequests();
+    }
+
+    processRequests() {
+        if (this.pending || this.queue.length === 0) {
+            return;
+        }
+        this.pending = true;
+        let dueForEvent = false;
+        if (!this.eventData.anonId || !this.eventData.lastSuccess) {
+            //Retrieve cached data
+            if (this.localStorageAvailable) {
+                try {
+                    const data = window.localStorage.getItem(this.STORAGE_TOKEN);
+                    if (data) {
+                        const json = JSON.parse(data);
+                        this.eventData.anonId = json.anonId;
+                        this.eventData.lastSuccess = Number(json.lastSuccess);
+                    }
+                } catch (e) {
+                    warnOnce('Unable to read from LocalStorage');
+                }
+            }
+        }
+
+        if (!validateUuid(this.eventData.anonId)) {
+            this.eventData.anonId = uuid();
+            dueForEvent = true;
+        }
+        const nextUpdate = this.queue.shift();
+
+        // Record turnstile event once per calendar day.
+        if (this.eventData.lastSuccess) {
+            const lastUpdate = new Date(this.eventData.lastSuccess);
+            const nextDate = new Date(nextUpdate);
+            const daysElapsed = (nextUpdate - this.eventData.lastSuccess) / (24 * 60 * 60 * 1000);
+            dueForEvent = dueForEvent || daysElapsed >= 1 || daysElapsed < 0 || lastUpdate.getDate() !== nextDate.getDate();
```

Comment:
What's the reason for the `lastUpdate.getDate() !== nextDate.getDate()` condition?  When _would_ these two dates be equal?

Reply:
`daysElapsed` is not perfectly named. It calculates how many 24-hr periods have elapsed since the last update. Notifications should be sent once per calendar day, so the date check allows testing for situations where a user loads the map at 11pm and again at 1am. 

`Date#getDate()` returns just the day of the month, so it would be equal on 6/20 and 7/20. The combined check returns true if the clock has gone backwards (fly east), or forward ( by 24+ hrs or by a calendar date).